File: backend/app/games/solitaire/engine.py
"""Le Klondike (pioche une carte, talon repassé sans limite), pour vérifier une partie.

Le joueur joue en local ; le serveur ne voit la partie qu'à la fin, sous forme de la
liste des coups, qu'il rejoue ici sur la donne qu'il a tirée. Le même moteur existe
côté client (frontend/src/games/solitaire/engine.ts) : les deux doivent rester
identiques, coup pour coup.

Cartes : rang (A23456789TJQK) + couleur (h d c s), « Th » = dix de cœur. Une donne est
la suite des 52 codes collés (104 caractères).

Coups :
- « d » : retourner la carte du talon sur la défausse ; talon vide, la défausse
  repasse en talon ;
- « <source>><cible> » ou « <source>><cible>:<n> » : déplacer la carte du dessus (ou
  les n du dessus d'une colonne). Piles : « w » défausse, « t0 »…« t6 » colonnes,
  « f0 »…« f3 » fondations.

La carte cachée découverte par un déplacement se retourne d'elle-même.
"""

from __future__ import annotations

import secrets
from dataclasses import dataclass, field

from app.games.base import GameError

RANKS = "A23456789TJQK"
SUITS = "hdcs"
RED = {"h", "d"}
COLUMNS = 7
FOUNDATIONS = 4
MAX_MOVES = 5000


def new_deck() -> str:
    """Un paquet battu au hasard pur (secrets), rien de choisi ni de filtré."""
    cards = [r + s for s in SUITS for r in RANKS]
    for i in range(len(cards) - 1, 0, -1):
        j = secrets.randbelow(i + 1)
        cards[i], cards[j] = cards[j], cards[i]
    return "".join(cards)


def parse_deck(deck: str) -> list[str]:
    cards = [deck[i : i + 2] for i in range(0, len(deck), 2)]
    if len(cards) != 52 or len(set(cards)) != 52:
        raise ValueError("Donne invalide.")
    return cards


def rank(card: str) -> int:
    return RANKS.index(card[0]) + 1


def red(card: str) -> bool:
    return card[1] in RED


@dataclass
class Klondike:
    # Colonnes : (carte, visible), du fond vers le dessus.
    tableau: list[list[tuple[str, bool]]] = field(default_factory=list)
    # Talon et défausse : le dessus est la fin de la liste.
    stock: list[str] = field(default_factory=list)
    waste: list[str] = field(default_factory=list)
    foundations: list[list[str]] = field(default_factory=list)

    @classmethod
    def deal(cls, deck: str) -> Klondike:
        """Distribution à la main, rangée par rangée : la colonne i reçoit i+1 cartes,
        la dernière face visible. Les 24 restantes font le talon."""
        cards = parse_deck(deck)
        tableau: list[list[tuple[str, bool]]] = [[] for _ in range(COLUMNS)]
        k = 0
        for row in range(COLUMNS):
            for col in range(row, COLUMNS):
                tableau[col].append((cards[k], row == col))
                k += 1
        return cls(
            tableau=tableau,
            stock=cards[k:],
            waste=[],
            foundations=[[] for _ in range(FOUNDATIONS)],
        )

    def won(self) -> bool:
        return all(len(f) == 13 for f in self.foundations)

    # --- Coups -----------------------------------------------------------------

    def play(self, move: str) -> None:
        if move == "d":
            self._draw()
            return
        try:
            src, rest = move.split(">", 1)
            dst, _, count = rest.partition(":")
            n = int(count) if count else 1
        except ValueError:
            raise GameError("Coup illisible.") from None
        # D'une pile vers elle-même, ou d'une fondation à l'autre : ce n'est pas un coup.
        if src == dst or (src[:1] == "f" and dst[:1] == "f"):
            raise GameError("Cette carte ne va pas là.")
        cards = self._take(src, n)
        self._put(dst, cards)
        self._pop(src, n)

    def _draw(self) -> None:
        if self.stock:
            self.waste.append(self.stock.pop())
        elif self.waste:
            self.stock = self.waste[::-1]
            self.waste = []
        else:
            raise GameError("Le talon est vide.")

    def _pile(self, name: str) -> tuple[str, int]:
        if name == "w":
            return "w", 0
        if len(name) == 2 and name[0] in "tf" and name[1].isdigit():
            index = int(name[1])
            if index < (COLUMNS if name[0] == "t" else FOUNDATIONS):
                return name[0], index
        raise GameError("Pile inconnue.")

    def _take(self, src: str, n: int) -> list[str]:
        """Les n cartes du dessus de la source, dans l'ordre (sans les retirer)."""
        kind, index = self._pile(src)
        if kind == "t":
            column = self.tableau[index]
            if n < 1 or n > len(column) or not all(up for _, up in column[-n:]):
                raise GameError("Ces cartes ne se déplacent pas.")
            return [card for card, _ in column[-n:]]
        pile = self.waste if kind == "w" else self.foundations[index]
        if n != 1 or not pile:
            raise GameError("Ces cartes ne se déplacent pas.")
        return [pile[-1]]

    def _put(self, dst: str, cards: list[str]) -> None:
        kind, index = self._pile(dst)
        first = cards[0]
        if kind == "f":
            pile = self.foundations[index]
            fits = len(cards) == 1 and (
                (not pile and rank(first) == 1)
                or (pile and pile[-1][1] == first[1] and rank(first) == rank(pile[-1]) + 1)
            )
            if not fits:
                raise GameError("Cette carte ne va pas là.")
            pile.append(first)
            return
        if kind != "t":
            raise GameError("Cette carte ne va pas là.")
        column = self.tableau[index]
        # Une suite déplacée est toujours alternée et descendante : elle n'a pu se
        # former qu'ainsi. On le revérifie quand même, le client n'est pas juge.
        for upper, lower in zip(cards, cards[1:], strict=False):
            if red(upper) == red(lower) or rank(lower) != rank(upper) - 1:
                raise GameError("Cette carte ne va pas là.")
        if column:
            top, up = column[-1]
            fits = up and red(top) != red(first) and rank(first) == rank(top) - 1
        else:
            fits = rank(first) == 13
        if not fits:
            raise GameError("Cette carte ne va pas là.")
        column.extend((card, True) for card in cards)

    def _pop(self, src: str, n: int) -> None:
        """Retire les cartes parties de la source ; retourne la carte découverte."""
        kind, index = self._pile(src)
        if kind == "t":
            column = self.tableau[index]
            del column[-n:]
            if column and not column[-1][1]:
                column[-1] = (column[-1][0], True)
        elif kind == "w":
            self.waste.pop()
        else:
            self.foundations[index].pop()


def replay(deck: str, moves: list[str]) -> Klondike:
    """Rejoue la partie coup par coup ; GameError au premier coup illégal."""
    if len(moves) > MAX_MOVES:
        raise GameError("Partie trop longue.")
    game = Klondike.deal(deck)
    for move in moves:
        game.play(move)
    return game

File: frontend/src/games/solitaire/engine.ts
/* Le Klondike (pioche une carte, talon repassé sans limite). Miroir exact de
   backend/app/games/solitaire/engine.py : le serveur rejoue la liste des coups avec son
   propre moteur pour valider la victoire, les deux doivent accepter les mêmes coups.

   Cartes : rang (A23456789TJQK) + couleur (h d c s), « Th » = dix de cœur.
   Coups : « d » (piocher, ou retourner la défausse en talon), « src>dst » ou
   « src>dst:n » (les n cartes du dessus d'une colonne). Piles : « w » défausse,
   « t0 »…« t6 » colonnes, « f0 »…« f3 » fondations. */

export type Card = string;
export type Slot = { card: Card; up: boolean };

export type State = {
  tableau: Slot[][];
  /* Talon et défausse : le dessus est la fin du tableau. */
  stock: Card[];
  waste: Card[];
  foundations: Card[][];
};

export const RANKS = "A23456789TJQK";
export const COLUMNS = 7;
export const FOUNDATIONS = 4;

export const rank = (card: Card) => RANKS.indexOf(card[0]) + 1;
export const suit = (card: Card) => card[1] as "h" | "d" | "c" | "s";
export const isRed = (card: Card) => card[1] === "h" || card[1] === "d";

export function parseDeck(deck: string): Card[] {
  const cards: Card[] = [];
  for (let i = 0; i < deck.length; i += 2) cards.push(deck.slice(i, i + 2));
  return cards;
}

/* Rangée par rangée, comme à la main : la colonne i reçoit i+1 cartes, la dernière
   visible. Les 24 restantes font le talon. */
export function deal(deck: string): State {
  const cards = parseDeck(deck);
  const tableau: Slot[][] = Array.from({ length: COLUMNS }, () => []);
  let k = 0;
  for (let row = 0; row < COLUMNS; row++) {
    for (let col = row; col < COLUMNS; col++) {
      tableau[col].push({ card: cards[k], up: row === col });
      k++;
    }
  }
  return {
    tableau,
    stock: cards.slice(k),
    waste: [],
    foundations: Array.from({ length: FOUNDATIONS }, () => []),
  };
}

/* L'ordre de distribution (colonne, place dans la colonne), pour l'animer. */
export const DEAL_ORDER: { col: number; row: number }[] = (() => {
  const order = [];
  for (let row = 0; row < COLUMNS; row++)
    for (let col = row; col < COLUMNS; col++) order.push({ col, row });
  return order;
})();

export function isWon(state: State): boolean {
  return state.foundations.every((f) => f.length === 13);
}

function clone(state: State): State {
  return {
    tableau: state.tableau.map((c) => c.slice()),
    stock: state.stock.slice(),
    waste: state.waste.slice(),
    foundations: state.foundations.map((f) => f.slice()),
  };
}

type Pile = { kind: "w" | "t" | "f"; index: number };

function pile(name: string): Pile | null {
  if (name === "w") return { kind: "w", index: 0 };
  if (/^[tf]\d$/.test(name)) {
    const index = Number(name[1]);
    if (index < (name[0] === "t" ? COLUMNS : FOUNDATIONS))
      return { kind: name[0] as "t" | "f", index };
  }
  return null;
}

/* Les cartes du dessus de la source, dans l'ordre, sans les retirer. */
function take(state: State, src: Pile, n: number): Card[] | null {
  if (src.kind === "t") {
    const column = state.tableau[src.index];
    if (n < 1 || n > column.length) return null;
    const moved = column.slice(-n);
    return moved.every((s) => s.up) ? moved.map((s) => s.card) : null;
  }
  const from = src.kind === "w" ? state.waste : state.foundations[src.index];
  return n === 1 && from.length ? [from[from.length - 1]] : null;
}

export function fitsFoundation(foundation: Card[], card: Card): boolean {
  if (!foundation.length) return rank(card) === 1;
  const top = foundation[foundation.length - 1];
  return suit(top) === suit(card) && rank(card) === rank(top) + 1;
}

export function fitsColumn(column: Slot[], card: Card): boolean {
  if (!column.length) return rank(card) === 13;
  const top = column[column.length - 1];
  return top.up && isRed(top.card) !== isRed(card) && rank(card) === rank(top.card) - 1;
}

/* Le coup appliqué (nouvel état), ou null s'il est illégal. */
export function apply(state: State, move: string): State | null {
  if (move === "d") {
    const next = clone(state);
    if (next.stock.length) next.waste.push(next.stock.pop()!);
    else if (next.waste.length) {
      next.stock = next.waste.reverse();
      next.waste = [];
    } else return null;
    return next;
  }
  const match = /^(\w+)>(\w+)(?::(\d+))?$/.exec(move);
  if (!match) return null;
  const [, srcName, dstName, count] = match;
  if (srcName === dstName || (srcName[0] === "f" && dstName[0] === "f")) return null;
  const src = pile(srcName);
  const dst = pile(dstName);
  if (!src || !dst) return null;
  const n = count ? Number(count) : 1;
  const cards = take(state, src, n);
  if (!cards) return null;

  if (dst.kind === "f") {
    if (cards.length !== 1 || !fitsFoundation(state.foundations[dst.index], cards[0])) return null;
  } else if (dst.kind === "t") {
    for (let i = 1; i < cards.length; i++) {
      if (isRed(cards[i]) === isRed(cards[i - 1]) || rank(cards[i]) !== rank(cards[i - 1]) - 1)
        return null;
    }
    if (!fitsColumn(state.tableau[dst.index], cards[0])) return null;
  } else return null;

  const next = clone(state);
  if (dst.kind === "f") next.foundations[dst.index].push(cards[0]);
  else next.tableau[dst.index].push(...cards.map((card) => ({ card, up: true })));

  if (src.kind === "t") {
    const column = next.tableau[src.index];
    column.splice(column.length - n, n);
    const top = column[column.length - 1];
    if (top && !top.up) column[column.length - 1] = { card: top.card, up: true };
  } else if (src.kind === "w") next.waste.pop();
  else next.foundations[src.index].pop();
  return next;
}

/* Où est une carte : sa pile et sa place dans la pile. */
export type Where = { pile: string; index: number };

export function locate(state: State, card: Card): Where | null {
  for (let c = 0; c < COLUMNS; c++) {
    const i = state.tableau[c].findIndex((s) => s.card === card);
    if (i >= 0) return { pile: `t${c}`, index: i };
  }
  for (let f = 0; f < FOUNDATIONS; f++) {
    const i = state.foundations[f].indexOf(card);
    if (i >= 0) return { pile: `f${f}`, index: i };
  }
  let i = state.waste.indexOf(card);
  if (i >= 0) return { pile: "w", index: i };
  i = state.stock.indexOf(card);
  if (i >= 0) return { pile: "s", index: i };
  return null;
}

/* Le nombre de cartes qu'on emporte en saisissant cette carte (elle et celles posées
   dessus), 0 si elle ne se saisit pas. */
export function grabCount(state: State, where: Where): number {
  if (where.pile === "s") return 0;
  if (where.pile === "w") return where.index === state.waste.length - 1 ? 1 : 0;
  if (where.pile[0] === "f") {
    const f = state.foundations[Number(where.pile[1])];
    return where.index === f.length - 1 ? 1 : 0;
  }
  const column = state.tableau[Number(where.pile[1])];
  return column[where.index].up ? column.length - where.index : 0;
}

export function moveString(src: string, dst: string, n: number): string {
  return n > 1 ? `${src}>${dst}:${n}` : `${src}>${dst}`;
}

/* Le tap : la carte part d'elle-même au meilleur endroit. Fondation d'abord (si c'est
   la carte du dessus), sinon une colonne où elle se pose ; un roi ne quitte pas le fond
   d'une colonne pour une autre colonne vide. */
export function tapMove(state: State, where: Where): string | null {
  const n = grabCount(state, where);
  if (!n) return null;
  const src = where.pile;
  const card =
    src === "w"
      ? state.waste[where.index]
      : src[0] === "f"
        ? state.foundations[Number(src[1])][where.index]
        : state.tableau[Number(src[1])][where.index].card;

  if (n === 1 && src[0] !== "f") {
    // La fondation de sa couleur, sinon (as) la première vide.
    let target = state.foundations.findIndex(
      (f) => f.length > 0 && fitsFoundation(f, card),
    );
    if (target < 0 && rank(card) === 1) target = state.foundations.findIndex((f) => !f.length);
    if (target >= 0) return moveString(src, `f${target}`, 1);
  }

  const kingAtBottom = src[0] === "t" && where.index === 0;
  let empty: number | null = null;
  for (let c = 0; c < COLUMNS; c++) {
    if (`t${c}` === src) continue;
    const column = state.tableau[c];
    if (!column.length) {
      if (empty === null && rank(card) === 13 && !kingAtBottom) empty = c;
      continue;
    }
    if (fitsColumn(column, card)) return moveString(src, `t${c}`, n);
  }
  return empty !== null ? moveString(src, `t${empty}`, n) : null;
}

/* Plus aucune carte cachée dans les colonnes : la partie est gagnée d'avance (on
   peut toujours jouer la plus petite carte restante). */
export function allRevealed(state: State): boolean {
  return state.tableau.every((column) => column.every((s) => s.up));
}

/* Les coups qui finissent une partie gagnée d'avance : à chaque fois une carte du plus
   petit rang restant (elle va forcément sur sa fondation), en piochant au talon tant
   qu'aucune n'est à portée. */
export function finishingMoves(state: State): string[] {
  const moves: string[] = [];
  let current = state;
  const play = (move: string) => {
    current = apply(current, move)!;
    moves.push(move);
  };
  while (!isWon(current)) {
    const lowest = Math.min(
      ...current.tableau.flatMap((column) => column.map((s) => rank(s.card))),
      ...current.stock.map(rank),
      ...current.waste.map(rank),
    );
    const reachable: { card: Card; src: string }[] = [];
    current.tableau.forEach((column, c) => {
      if (column.length) reachable.push({ card: column[column.length - 1].card, src: `t${c}` });
    });
    if (current.waste.length)
      reachable.push({ card: current.waste[current.waste.length - 1], src: "w" });
    const next = reachable.find((r) => rank(r.card) === lowest);
    if (next) {
      const target = current.foundations.findIndex((f) => fitsFoundation(f, next.card));
      play(`${next.src}>f${target}`);
    } else play("d");
  }
  return moves;
}
